src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+package org.ohdsi.webapi.shiro.filters;
+
+import java.io.IOException;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.shiro.SecurityUtils;
+import org.apache.shiro.web.servlet.AdviceFilter;
+import org.apache.shiro.web.util.WebUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.ohdsi.webapi.shiro.PermissionManager;
+import org.ohdsi.webapi.shiro.Entities.RoleEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * Filter class to dynamically assign a "team project" role to the
+ * user, after validating with an external authorization service
+ * whether the user has been granted access to this "team project".
+ * 
+ * @author Pieter Lukasse
+ */
+public class TeamProjectBasedAuthorizingFilter extends AdviceFilter {
+
+  private final Logger logger = LoggerFactory.getLogger(TeamProjectBasedAuthorizingFilter.class);
+
+  private final PermissionManager authorizer;
+  private final Set<String> defaultRoles;
+  private final String authorizationMode;
+  private final String authorizationUrl;
+
+  public TeamProjectBasedAuthorizingFilter(
+          PermissionManager authorizer,
+          Set<String> defaultRoles,
+          String authorizationMode,
+          String authorizationUrl) {
+    this.authorizer = authorizer;
+    this.defaultRoles = defaultRoles;
+    this.authorizationMode = authorizationMode;
+    this.authorizationUrl = authorizationUrl;
+    logger.debug("AUTHORIZATION_MODE in TeamProjectBasedAuthorizingFilter constructor == '{}'", this.authorizationMode);
+    logger.debug("AUTHORIZATION_URL in TeamProjectBasedAuthorizingFilter constructor == '{}'", this.authorizationUrl);
+  }
+
+  @Override
+  protected boolean preHandle(ServletRequest request, ServletResponse response) throws Exception {
+
+    try {
+        logger.debug("preHandle in TeamProjectBasedAuthorizingFilter == '{}'", this.authorizationMode);
+        if (this.authorizationMode.equals("teamproject") && SecurityUtils.getSubject().isAuthenticated()) {
+          // in case of "teamproject" mode, we want all roles to be reset always, and
+          // set to only the one requested/found in the request parameters (following lines below):
+          String login = this.authorizer.getCurrentUser().getLogin();
+          boolean foundValidTeamProject = extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(this, login, request, response);
```

Comment:
style: probably could remove the "IfNecessary" from the method name to shorten it (optional)

Reply:
I like the longer name, since otherwise it sounds like the method is always updating the user...I'll leave it for now and see if others complain. Perhaps the method should just be broken up into smaller ones that do just one thing... leaving for later, and I'll adjust if necessary.